aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -0,0 +1,514 @@
+#include "TensorIterator.h"
+
+#include <ATen/ExpandUtils.h>
+#include <ATen/Parallel.h>
+#include <time.h>
+
+namespace at {
+
+void TensorIterator::reorder_dimensions() {
+  // Sort the dimensions based on the sum-of-strides in ascending order. NOTE:
+  // that this inverts the order of C-contiguous tensors. strides[0] is the
+  // fastest moving dimension instead of strides[ndim - 1].
+
+  auto sum_of_strides = SmallVector<double, 6>(ndim(), 0.0);
+  for (int dim = 0; dim < ndim(); dim++) {
+    double sum = 0.0;
+    for (const auto& op : operands_) {
+      if (op.stride_.size() == 0) continue;
+      sum += op.stride_[dim];
+    }
+
+    // Weight each dimension by its index. Given two dimensions with equal
+    // some of strides, this preserves the given relative ordering.
+    sum += (ndim() - dim - 1) / (double)ndim();
+
+    sum_of_strides[dim] = sum;
+  }
+
+  // initialize perm with 0, 1, 2, ...
+  perm_.resize(ndim());
+  std::iota(std::begin(perm_), std::end(perm_), 0);
+
+  std::sort(std::begin(perm_), std::end(perm_), [&](size_t i1, size_t i2) {
+    return sum_of_strides[i1] < sum_of_strides[i2];
+  });
+
+  auto reorder = [](IntList data, IntList perm_) {
+    auto res = DimVector(data.size(), 0);
+    for (size_t i = 0; i < perm_.size(); i++) {
+      res[i] = data[perm_[i]];
+    }
+    return res;
+  };
+
+  // Update shape and strides
+  shape_ = reorder(shape_, perm_);
+  for (auto& op : operands_) {
+    if (op.stride_.size() > 0) {
+      op.stride_ = reorder(op.stride_, perm_);
+    }
+  }
+}
+
+template <typename F>
+static std::tuple<ScalarType, Backend>
+compute_result_type(at::ArrayRef<OperandInfo> operands, const F& predicate) {
+  auto result_type = ScalarType::Undefined;
+  auto backend = Backend::Undefined;
+  for (auto& op : operands) {
+    if (!op.tensor_->defined()) continue;
+    if (!predicate(*op.tensor_)) continue;
+    auto dtype = op.tensor_->type().scalarType();;
+    result_type = (result_type == ScalarType::Undefined
+        ? dtype
+        : promoteTypes(result_type, dtype));
+    backend = (backend == Backend::Undefined
+        ? op.tensor_->type().backend()
+        : backend);
+  }
+  return std::make_tuple(result_type, backend);
+}
+
+void TensorIterator::compute_common_type() {
+  // The result dtype is computed with the precedence:
+  // 1) Tensors of rank one or higher
+  // 2) Tensors of any rank that aren't wrapped numbers
+  // 3) Any tensor
+  //
+  // For example:
+  //  torch.randn(3, 3) + torch.tensor([1.0], dtype=torch.double) -> torch.double
```

Comment:
the LHS here and below don't specify a dtype (it depends on what the default is set to).